### PR TITLE
F/npcs

### LINKED ISSUE #32: NPC Animations and Steering Behaviors
* Create a few different NPC prefabs using the models on Mixamo (Around 3?)
 * Create an animation controller for each NPC prefab (See player animation controller to replicate) 
 * Add basic steering behaviors for NPCs, such as Arrive/Wander/Idle (Like in assignment 1)

### PR DESCRIPTION
**Pull request by Aleoris**
- Added 4 NPC prefabs
- They are currently in the in just for display and testing purposes
- They have 3 states (chase, wander, idle)
- When bumping into them (animation + sound)
- animator copied from player

### CODE REVIEW

File: Assets/Scripts/NPCs/StateMachine/States/Wander.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+        if (numberOfMoves == 0)
+        {
+            StopMoving(npc);
+        }
+    }
+
+    private void FindNewTarget(GameObject npc)
+    {
+        Vector3 npcVelocity = npc.GetComponent<NPCManager>().Velocity;
```
Comment (tedgag) [inline Assets/Scripts/NPCs/StateMachine/States/Wander.cs]: I wonder if we could make move all the GetComponent() calls outside of this class and store the values in variables, since Rider seem to think that they are expensive method evocation. I saw that in NPCStateMachine we are already storing a NPCManager, so maybe instead of parsing the npc here we could make the GetComponent call in NPCManager and pass the velocity as an argument?

File: Assets/Scripts/NPCs/StateMachine/States/Wander.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        Vector3 direction = Vector3.Normalize(_target - npc.transform.position);
+
+        //The NPC will always looking where it needs to go.
+        LookWhereYouAreGoing(npc, direction);
+
+        //The NPC always walk forward.
+        Vector3 velocity = Vector3.forward * NPCsGlobalVariables.WanderMaxVelocity;
+
+        npc.GetComponent<Animator>().SetFloat(NPCsGlobalVariables.VelocityXHash, velocity.x);
```
Comment (tedgag) [inline Assets/Scripts/NPCs/StateMachine/States/Wander.cs]: Same as above, we could make this method return this velocity and set the animator values in NPCManager instead, and store the animator as a variable there.

File: Assets/Scripts/Player/PlayerDataManager.cs [hunk 3]
```diff
@@ -7,9 +7,21 @@ public class PlayerDataManager : MonoBehaviour
...
+        private void Start()
         {
             PlayerData = new PlayerData(gameObject.transform.position);
         }
+        
+        /*
+         * Since the rigidbody is not directly on the game object that is moving. We cannot use it to get the velocity.
+         * The velocity needs to be computed manually.
+         */
+        private void Update()
```
Comment (tedgag) [inline Assets/Scripts/Player/PlayerDataManager.cs]: Might not be needed later since I found a fix to having the rigid body attached to the player directly in my branch, but it works for now, so we should refactor this later

File: Assets/Scripts/NPCs/StateMachine/States/Chase.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        Chasing(player, npc);
+    }
+    
+    private void Chasing(GameObject player, GameObject npc)
+    {
+        Vector3 futureLocation = player.transform.position + player.GetComponent<PlayerDataManager>().PlayerData.Velocity * NPCsGlobalVariables.ChasePredictionMultiplier;
+        
+        Vector3 desiredVelocity = Vector3.Normalize(futureLocation - npc.transform.position) * NPCsGlobalVariables.ChaseAcceleration;
+        
+        Vector3 currentVelocity = npc.GetComponent<NPCManager>().Velocity;
```
Comment (tedgag) [inline Assets/Scripts/NPCs/StateMachine/States/Chase.cs]: See comments for the Wander file, the same idea could be applied here

File: Assets/Scripts/NPCs/NPCsGlobalVariables.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+﻿using UnityEngine;
+public class NPCsGlobalVariables
```
Comment (tedgag) [inline Assets/Scripts/NPCs/NPCsGlobalVariables.cs]: All these variables could be stored in Indestructibles.cs, where we store all our other global variables. It would be consistent with the rest of the code and we wouldn't need this extra file

**Review by tedgag — CHANGES_REQUESTED** [review]
Looks great! I like the animation that you chose for bumping into NPCs, andfor the most part the steering behaviors work as expected. One thing that i would add to the steering would be the idea of a FOV radius for the NPC like in assigmnent 1 where, if their target is not in their FOV radius, they need to stop and rotate toward their target without moving, and set off once there target is in their FOV. A side-step radius like in assignment 1 might be nice as well so the NPC does not have to rotate toward the player to move toward them. Wander might need some refactoring as well when we have the navmesh so the NPC do not choose wander target inside the environment objects.

File: Assets/Animations/player/player_animation_controller.controller [hunk 6]
```diff
@@ -28,119 +22,6 @@ AnimatorStateMachine:
   m_ExitPosition: {x: 500, y: 70, z: 0}
   m_ParentStateMachinePosition: {x: 800, y: 20, z: 0}
   m_DefaultState: {fileID: 8334071979550753427}
---- !u!206 &-4180541552395086271
-BlendTree:
-  m_ObjectHideFlags: 1
```
Comment (f-michaut) [inline Assets/Animations/player/player_animation_controller.controller]: Why did you remove the player's animator blend tree ?

File: Assets/Scripts/NPCs/NPCsGlobalVariables.cs (hunk 5 above)
Comment (f-michaut) [reply Assets/Scripts/NPCs/NPCsGlobalVariables.cs]: Having different specialised global class could be usefull, because all those globals don't concern the same things
Thoses variables are only for the NPCs, having them in a separate file seems appropriate and avoid putting everything in the same file
This is good for me

File: Assets/Scripts/NPCs/StateMachine/States/State.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿using UnityEngine;
+
+public abstract class State
+{
+    protected string name;
+
+    public abstract void Move(GameObject player, GameObject npc);
+
+    public string Name => name;
```
Comment (f-michaut) [inline Assets/Scripts/NPCs/StateMachine/States/State.cs]: What is the Name used for exactly ?

File: Assets/Scripts/NPCs/StateMachine/States/Chase.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public override void Move(GameObject player, GameObject npc)
+    {
+        Chasing(player, npc);
+    }
+    
+    private void Chasing(GameObject player, GameObject npc)
+    {
+        Vector3 futureLocation = player.transform.position + player.GetComponent<PlayerDataManager>().PlayerData.Velocity * NPCsGlobalVariables.ChasePredictionMultiplier;
```
Comment (f-michaut) [inline Assets/Scripts/NPCs/StateMachine/States/Chase.cs]: Avoid calling GetComponent each time this function is called
Instead create a private PlayerDataManager variable in the class (or the base class, up to you) like this : 
```c#
    private PlayerDataManager _manager = player.GetComponent<PlayerDataManager>();
```

This way, the call will be done only once at the initialization, and will be more efficient
(do this for all the GetComponents call, a GetComponent call is expensive and will always return the same result, no need to call it everytime we need the same component)

File: Assets/Sound/NPCs/DC_Motors (1).pdf.meta [hunk 9]
```diff
@@ -0,0 +1,7 @@
+fileFormatVersion: 2
+guid: 24af55e6ef628964580a22aaeb6b40e2
+DefaultImporter:
```
Comment (f-michaut) [inline Assets/Sound/NPCs/DC_Motors (1).pdf.meta]: Can you remove thoses PDF files ?

**Review by f-michaut — CHANGES_REQUESTED** [review]

File: Assets/Sound/NPCs/DC_Motors (1).pdf.meta (hunk 9 above)
Comment (Aleoris) [reply Assets/Sound/NPCs/DC_Motors (1).pdf.meta]: Oops xD

File: Assets/Scripts/NPCs/StateMachine/States/Wander.cs (hunk 1 above)
Comment (Aleoris) [reply Assets/Scripts/NPCs/StateMachine/States/Wander.cs]: Yeah, I forgot to move them into variables when I was cleaning the code. Will do!

File: Assets/Scripts/NPCs/StateMachine/States/Chase.cs (hunk 8 above)
Comment (Aleoris) [reply Assets/Scripts/NPCs/StateMachine/States/Chase.cs]: Yeah, I forgot to move them into variables when I was cleaning the code. Will do!

File: Assets/Scripts/NPCs/StateMachine/States/State.cs (hunk 7 above)
Comment (Aleoris) [reply Assets/Scripts/NPCs/StateMachine/States/State.cs]: I was using that variable at some point but forgot to remove it.

File: Assets/Animations/player/player_animation_controller.controller (hunk 6 above)
Comment (Aleoris) [reply Assets/Animations/player/player_animation_controller.controller]: There were 2 useless states (Idle and BlendTree) that I removed. The BlendTree is still there, I just removed an empty state called BlendTree.

File: Assets/Scripts/NPCs/StateMachine/States/Wander.cs (hunk 2 above)
Comment (Aleoris) [reply Assets/Scripts/NPCs/StateMachine/States/Wander.cs:74]: Done

File: Assets/Scripts/Player/PlayerDataManager.cs (hunk 3 above)
Comment (Aleoris) [reply Assets/Scripts/Player/PlayerDataManager.cs]: Yup, thats why I wanted your PR to be merged so I can adapt my code.

**Review by tedgag — COMMENTED** [review]
With the FOV it looks much better to me, I only have a small comment about the functionality. I noticed that when you're being chased and you run around in tight circles, the NPCs look a bit weird since they stop moving to rotate, then move toward player, then stop moving etc. I think if we don't have the side step radius we should have a small satisfaction radius that when the NPC to their target is close enough they will stop moving and rotating. I also think this might be caused by the fact that the acceleration might a bit too high, since NPC tend to get to their max speed from rest too quickly. lowering it would make it look less weird when move back and forth from moving to not moving.

**Review by tedgag — APPROVED** [review]
The changes in acceleration and the satisfaction radius makes it way smoother! I don't have any more comments for this, good job!

File: Assets/Scripts/NPCs/StateMachine/NPCStateMachine.cs [hunk 10]
```diff
@@ -40,10 +37,11 @@ private void Update()
...
+     * For now, the collider is of type trigger because there is no navmesh.
      */
-    public void CollidedWithPlayer()
+    public void OnTriggerEnter(Collider collider)
     {
-        if (_currentState != _chase)
+        if (_currentState != _chase && collider.gameObject.CompareTag("Player"))
         {
             GetComponent<AudioSource>().PlayOneShot(bumpingSounds[0]);
             GetComponent<Animator>().SetTrigger(Animator.StringToHash("Get Hit"));
```
Comment (f-michaut) [inline Assets/Scripts/NPCs/StateMachine/NPCStateMachine.cs]: Thoses `GetComponent` calls will be repeated each time, move them to a class variable

**Review by f-michaut — CHANGES_REQUESTED** [review]
Great work, I love the sounds !
Just a small fix on GetComponents and it should be good !

Also, can you make NPCs collide with the environment ?
And the player is missing the "player" tag, can you fix that ?

File: Assets/Scripts/NPCs/StateMachine/NPCStateMachine.cs (hunk 10 above)
Comment (Aleoris) [reply Assets/Scripts/NPCs/StateMachine/NPCStateMachine.cs]: Those are only called once per NPC. I guess I can put them in variable anyway.

Comment (f-michaut) [reply Assets/Scripts/NPCs/StateMachine/NPCStateMachine.cs]: Oh that's true, I forgot the if condition

**Review by f-michaut — APPROVED** [review]
